platform/web/request.go

```diff
@@ -71,91 +95,52 @@ func DecodeWithJSONSchema(request *http.Request, model interface{}, filePath str
-		}
 		return err
 	}
-	request.Body = io.NopCloser(bytes.NewBuffer(body))
-	payload := gojsonschema.NewBytesLoader(body)
-
-	result, err := gojsonschema.Validate(jsonSchema, payload)
-	if err != nil {
-		return fmt.Errorf("%w: %v", NewErrBadRequestResponse(nil), err)
+	if err := request.Body.Close(); err != nil {
+		return fmt.Errorf("cannot close request body: %v", err)
 	}
 
-	if !result.Valid() {
-		additionalPropertyErrors := make(ErrorDetails)
-		requiredPropertyErrors := make(ErrorDetails)
-		requestContentErrorDetails := make(ErrorDetails)
-
-		for _, desc := range result.Errors() {
-			propertyName := desc.Field()
-			if propertyName == "(root)" {
-				details, ok := desc.Details()["property"]
-				if ok {
-					propertyName = fmt.Sprintf("%v", details)
-				}
-			}
-
-			switch desc.Type() {
-			case "additional_property_not_allowed":
-				additionalPropertyErrors[propertyName] = desc.Description()
-			case "required":
-				requiredPropertyErrors[propertyName] = desc.Description()
-			default:
-				requestContentErrorDetails[propertyName] = desc.Description()
-			}
+	if err := validateJSONAgainstSchema(body, jsonSchema); err != nil {
+		var errMissingProperties JSONSchemaMissingPropertyError
+		if errors.As(err, &errMissingProperties) {
+			return fmt.Errorf("json schema validation error: %w", NewErrBadRequestResponse(errMissingProperties.Details))
 		}
-
-		if len(additionalPropertyErrors) > 0 {
-			errDetails := additionalPropertyErrors
-			for property, errMessage := range requestContentErrorDetails {
-				errDetails[property] = errMessage
-			}
-
-			return fmt.Errorf("the request body contains unknown keys: %w", NewErrBadRequestResponse(errDetails))
+		var errAdditionalProperties JSONSchemaAdditionalPropertyError
+		if errors.As(err, &errAdditionalProperties) {
+			return fmt.Errorf("json schema validation error: %w", NewErrBadRequestResponse(errAdditionalProperties.Details))
 		}
-
-		if len(requiredPropertyErrors) > 0 {
-			errDetails := requiredPropertyErrors
-			for property, errMessage := range requestContentErrorDetails {
-				errDetails[property] = errMessage
-			}
-
-			return fmt.Errorf("the request body contains unknown keys: %w", NewErrBadRequestResponse(errDetails))
+		var errSchemaValidation JSONSchemaValidationError
+		if errors.As(err, &errSchemaValidation) {
+			return fmt.Errorf("json schema validation error: %w", NewErrInvalidRequestBodyContent(errSchemaValidation.Details))
 		}
-
-		return fmt.Errorf("%w", NewErrInvalidRequestBodyContent(requestContentErrorDetails))
+		return fmt.Errorf("json schema validation error: %w", NewErrBadRequestResponse(nil))
 	}
 
-	if err := decode(request, model, options); err != nil {
+	if err := decode(body, model, options); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func decode(r *http.Request, val interface{}, options DecoderOptions) error {
-	decoder := json.NewDecoder(r.Body)
+func decode(input []byte, val interface{}, options DecoderOptions) error {
```

Comment:
Why are we still decoding the input if it's already been parsed and validated against with the JSON schema?

Reply:
We still need to fill the struct with the json data after validation though.